### DISCUSSION TITLE
Realtime on expressjs server

### DISCUSSION THREAD

**Discussion by henrykorir** [supabase/supabase#1204] [Questions] [answered]
How can I set up realtime on expressjs server and let the results be availed on the front-end part?

**Comment by silentworks** [comment, marked as answer]
In express you would use the supabase api as normal then inside of your frontend you just setup realtime listeners on the tables you want to listen to changes on. So if you were adding new data to a `tasks` table you would do the following inside of express:

```js
await supabase.from('tasks').insert([
  { name: 'Add new task' }
])
```

then inside of your frontend you just do the following:

```js
const taskListener = supabase.from('tasks')
    .on('INSERT', (payload) => handleNewTask(payload.new))
    .subscribe()
```

So on the second line we are listening for an `INSERT` event to happen and in the callback we are passing the payload to a `handleNewTask` function which you would create. Remember to unsubscribe to the `taskListener` at some point later in the code.

You can take a look at the Slack clone example code to see some more realtime events https://github.com/supabase/supabase/blob/master/examples/nextjs-slack-clone/lib/Store.js and also take a look at the documentation site about realtime https://supabase.io/docs/reference/javascript/subscribe.

**Reply by henrykorir** [reply]
The reason I would like to have the realtime on the server is because I would like to hide the keys. @silentworks the solution above you've provided entails setting up the supabase client on the front-end thus exposing the url and the keys. 
Is there a way I would totally hide the keys from the front-end while using realtime?

**Reply by kiwicopple** [reply]
Hey @henrykorir - the client library is isomorphic, so @silentworks 's solution above should work within the express server too. Are you experiencing any problems?

**Reply by silentworks** [reply]
@henrykorir you don't need to worry about the `anon` `public` key being exposed, this is how the system is built, it doesn't mean that someone will have access to your database to do writes or the likes as long as you have row level security turned on. Also if you were to do the realtime stuff in express you wouldn't be getting the benefit of realtime on the frontend because you would have to create your own websocket server to pass the data over in realtime which would defeat the purpose of supabase realtime.

**Reply by henrykorir** [reply]
@silentworks  @kiwicopple The explanations you've given settles it. Thanks.

**Reply by bryan-vh** [reply]
@silentworks @kiwicopple to follow up on this reply, what if our client is a mobile app (using native iOS or Android and not Flutter), any insights on how to move forward there? Seems like setting up a websocket server in express to pass the data from supabase realtime to those clients would be the only feasible option, no?
